### PR TITLE
Process torsion profiles

### PR DESCRIPTION
**Pull request by mattwthompson**
- [x] Add `TorsionStore` for storing torsion profiles as a database (#42)
- [x] Add input models
- [x] Add output models
- [x] Add MM optimization about target torsions
  - [x] Yoink https://github.com/lilyminium/openff-strike-team/blob/a6ccd2821ed627064529f5c4a22b47c1fa36efe2/torsions/datasets/mm/minimize-torsion-constrained.py#L35-L106
- [x] Add example plotting script (#41)
- [x] Add tests
- [ ] Update documentation

### CODE REVIEW

**Comment by mattwthompson** [comment]
Something's wrong with unit-handling.

I'm taking the energies from QCArchive to be in atomic units, like optimization datasets are, and [converting them to kcal/mol just the same](https://github.com/openforcefield/yammbs/blob/aaf731e5da212ae284fd34a00d19ce23c1f968b2/yammbs/torsion/inputs.py#L62-L65). I'm [getting minimized energies out of OpenMM in kcal/mol as well](https://github.com/openforcefield/yammbs/blob/aaf731e5da212ae284fd34a00d19ce23c1f968b2/yammbs/torsion/_minimize.py#L171-L177).

When I put together plots using some variant of `run_torsion_comparisons.py`, I need to multiply the QM energies by 10 (a completely arbitrary number) for the profiles to vaguely match up:

image

One could imagine or re-generate the plot with another factor besides 10, but the general conclusion (MM energies are about an order of magnitude higher, but similar in shape, to QM profiles) would be the same. Intuitively the scale of the QM profiles (a few kcal/mol up to a few tens of kcal/mol) seems reasonable and the raw MM values (up to a few hundred kcal/mol) seem like the wrong ones. But this is the closest I've come to being a force field scientist fitting torsions, so I don't trust my intuitions here very much.

The fact that the shape of (most of) the MM torsion profiles line up with the QM data implies to me that (most of) the minimizations are behaving. But the way I'm getting energies out is so obviously correct, [or at least what I do elsewhere](https://github.com/openforcefield/openff-interchange/blob/df397f4bf3c18bbad8e92f5d9c10eda4029935cd/openff/interchange/drivers/openmm.py#L123-L124) to me that I think I'm going crazy if that's the problem:

```python
energy=simulation.context.getState(
            getEnergy=True,
        )
        .getPotentialEnergy()
        .value_in_unit(
            openmm.unit.kilocalorie_per_mole,
        ),
```

Is there some gotcha I'm unfamiliar with? I assumed that this was as simple as (molar) energy at a handful of grid points, so simply kcal/mol or kJ/mol.

**Comment by amcisaac** [comment]
When I've worked on these, I just plotted the QM and MM energy of the QM structure, so I didn't do any minimizations. However, the QM and MM relative energies were both on the order of 0 to 10 kcal/mol (sometimes a bit larger, but not regularly 300). And you're right that the QM should indeed be coming out of QCA in hartrees. I don't see any obvious issues with your code, I wonder if it's something to do with the constrained optimization raising the energy of the MM structures?

File: yammbs/_molecule.py [hunk 1]
```diff
@@ -39,3 +40,12 @@ def _molecule_with_conformer_from_smiles(
     molecule.add_conformer(Quantity(conformer, "angstrom"))
 
     return molecule
+
+
+@lru_cache
+def _smiles_to_inchi_key(smiles: str) -> str:
+    from openff.toolkit import Molecule
+
+    return Molecule.from_mapped_smiles(smiles, allow_undefined_stereo=True).to_inchi(
```
Comment (jthorton) [inline yammbs/_molecule.py:49]: Just a note that this is the inchi not the inchikey as the function name suggests.

File: yammbs/torsion/inputs.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                    id=record.id,
+                    mapped_smiles=molecule.to_smiles(
+                        mapped=True,
+                        isomeric=True,
+                        explicit_hydrogens=True,
+                    ),
+                    dihedral_indices=record.specification.keywords.dihedrals[0],  # assuming this is only ever 1-len?
+                    coordinates={
+                        grid_id[0]: optimization.final_molecule.geometry
+                        for grid_id, optimization in record.minimum_optimizations.items()
```
Comment (jthorton) [inline yammbs/torsion/inputs.py:60]: I would use final molecules on the record to make this more clear.

File: yammbs/torsion/inputs.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                    },
+                    energies={
+                        grid_id[0]: optimization.energies[-1] * hartree2kcalmol
+                        for grid_id, optimization in record.minimum_optimizations.items()
```
Comment (jthorton) [inline yammbs/torsion/inputs.py:64]: Maybe use the final energies on the record to make this easier.

File: yammbs/torsion/inputs.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+import qcelemental
+from openff.qcsubmit.results import TorsionDriveResultCollection
+from pydantic import Field
+
+from yammbs._base.array import Array
+from yammbs._base.base import ImmutableModel
+
+hartree2kcalmol = qcelemental.constants.hartree2kcalmol
+bohr2angstroms = qcelemental.constants.bohr2angstroms
```
Comment (jthorton) [inline yammbs/torsion/inputs.py:11]: This is not being used anywhere and is probably the issue.

File: yammbs/torsion/inputs.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    tag: str
+
+
+class TorsionProfile(ImmutableModel):
+    mapped_smiles: str
+    dihedral_indices: list[int]
+
+    # TODO: Should this store more information than just the grid points and
+    #       final geometries? i.e. each point is tagged with an ID in QCArchive
+    coordinates: dict[float, Array]
```
Comment (jthorton) [inline yammbs/torsion/inputs.py:24]: What units do you expect here?

File: yammbs/torsion/inputs.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            qm_torsions=[
+                QCArchiveTorsionProfile(
+                    id=record.id,
+                    mapped_smiles=molecule.to_smiles(
+                        mapped=True,
+                        isomeric=True,
+                        explicit_hydrogens=True,
+                    ),
+                    dihedral_indices=record.specification.keywords.dihedrals[0],  # assuming this is only ever 1-len?
+                    coordinates={
```
Comment (jthorton) [inline yammbs/torsion/inputs.py:58]: This is saving the geometry as bohr is that what you expect?

File: yammbs/torsion/_store.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+
+            store.store_torsion_record(torsion_record)
+
+            for angle in qm_torsion.coordinates:
+                qm_point_record = QMTorsionPointRecord(
+                    molecule_id=store.get_molecule_id_by_smiles(
+                        torsion_record.mapped_smiles,
+                    ),
+                    grid_id=angle,  # TODO: This needs to be a tuple later
+                    coordinates=qm_torsion.coordinates[angle],
```
Comment (jthorton) [inline yammbs/torsion/_store.py:202]: This is in bohr but I think QMTorsionPointRecord is expecting angstrom.

File: yammbs/torsion/inputs.py (hunk 6 above)
Comment (mattwthompson) [reply yammbs/torsion/inputs.py:58]: Nope! Good catch

File: yammbs/_molecule.py (hunk 1 above)
Comment (mattwthompson) [reply yammbs/_molecule.py:49]: Oops - I'm pretty sure it's what we'd want, though? i.e. the names of the functions/models/etc should be updated, but the method call `Molecule.to_inchi` is correct?

File: yammbs/torsion/inputs.py (hunk 5 above)
Comment (mattwthompson) [reply yammbs/torsion/inputs.py:24]: Trying to keep things internally in AKMA[^1], will document units in these models now 😅 


[^1]: though I prefer SI, I think the small molecule scientists like Angstroms)

**Comment by mattwthompson** [comment]
Thanks @amcisaac and @jthorton! Re-generating the same data with some fixes produces data which seems much more reasonable:

image

There are still some details to hash through but I at least think it's in the right ballpark now

* Some bad optimizations?
* Noise due to making all torsion profiles relative to (its) minimum energy?
* Cleaner defaults for plots
  * Don't draw lines across parts of the grid that are missing lots of data
  * Better default axes ranges
  * Maybe some lightly-colored grids

File: yammbs/_molecule.py (hunk 1 above)
Comment (jthorton) [reply yammbs/_molecule.py:49]: I think Inchi or inchi key is fine, I like to use inchi key in qcsubmit as its fixed length but they should give the same result when checking if two molecules are the same.

**Comment by mattwthompson** [comment]
Okay, I'm done messing with the plots for now. @ntBre noticed that the MM profiles were normalized to the (location of the) MM minima whereas the convention is to use the (location of the) QM minimum. The differences are fairly small: 

image

**Comment by mattwthompson** [comment]
I consider this complete if all rows in the CSVs produced from this run are non-zero: https://github.com/openforcefield/yammbs-dataset-submission/actions/runs/12361612232/job/34499304927

**Comment by mattwthompson** [comment]
This is precisely what I hoped to see - the `0.0` rows are only in the red, small representative sample bit here:

https://github.com/openforcefield/yammbs-dataset-submission/pull/13/commits/b5deede6ce88649efca59121bc47c193a2e49040#diff-db3b6996a30e2a55bc83efbea52e18dbd86f135f9860641f6eb6d12f36571d4eL254-R258

image

**Pull request merged**